feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/service/storage/impl/MetisStorageServiceImplTestUpgradeLocalAnswerPost.kt

```diff
@@ -0,0 +1,211 @@
+            host = host,
+            metisContext = metisContext,
+            post = basePostUpdated
+        )
+
+        // Called by SendConversationPostWorker
+        sut.upgradeClientSideAnswerPost(
+            host = host,
+            metisContext = metisContext,
+            clientSidePostId = answerClientPostId,
+            post = answerUpdated
+        )
+
+        // THEN: Content stays the same and the upgrade is successful
+        assertStoredContentIsTheSame()
+        assertUpgradeSuccessful()
+    }
+
+    @Test
+    fun testUpgradeClientSideAnswerPost2() = runTest {
+        // GIVEN: A post with a new only local answer post
+        setupPostWithLocalAnswer()
+
+        // WHEN: upgradeClientSideAnswerPost is called before updatePost.
+        updateAnswerPostWithServerId()
+
+        // Called by SendConversationPostWorker
+        sut.upgradeClientSideAnswerPost(
+            host = host,
+            metisContext = metisContext,
+            clientSidePostId = answerClientPostId,
+            post = answerUpdated
+        )
+
+        // Called by the WebSocket
+        sut.updatePost(
+            host = host,
+            metisContext = metisContext,
+            post = basePostUpdated
+        )
+
+        // THEN: Content stays the same and the upgrade is successful
+        assertStoredContentIsTheSame()
+        assertUpgradeSuccessful()
+    }
+
+    private suspend fun setupPostWithLocalAnswer() {
+        sut.insertOrUpdatePosts(
+            host = host,
+            metisContext = metisContext,
+            posts = listOf(basePost)
+        )
+        sut.insertClientSidePost(
+            host = host,
+            metisContext = metisContext,
+            clientSidePostId = answerClientPostId,
+            post = localAnswer
+        )
+    }
+
+
+    private fun updateAnswerPostWithServerId() {
+        val answerPojoUpdated = localAnswerPojo.copy(serverPostIdCache = localAnswerPojo.serverPostIdCache.copy(serverPostId = 1))
+        basePostUpdated = StandalonePost(basePostPojo, conversation)
+        answerUpdated = AnswerPost(answerPojoUpdated, basePostUpdated)
+        basePostUpdated = basePostUpdated.copy(answers = listOf(answerUpdated))
+    }
+
+    private suspend fun assertStoredContentIsTheSame() {
+        val posts = getStoredPosts()
+        assertEquals(1, posts.size)
+        assertEquals(basePostPojo.content, posts.first().content)
+        assertEquals(1, posts.first().answers.size)
+        assertEquals(localAnswerPojo.content, posts.first().answers.first().content)
+    }
+
+    private suspend fun assertUpgradeSuccessful() {
+        val posts = getStoredPosts()
+        assertEquals(answerUpdated.serverPostId, posts.first().answers.first().serverPostId)
+    }
+
+    private suspend fun getStoredPosts() = sut.getStoredPosts(
+        serverId = host,
+        metisContext = metisContext
+    ).loadAsList()
+
+    private suspend fun <T : Any>PagingSource<Int, T>.loadAsList(): List<T> {
```

Comment:
I am not sure if this function is implemented correctly, could you try adding the paging-testing library and replacing the implementation of this function with this? 

```kotlin
private suspend fun <T : Any> PagingSource<Int, T>.loadAsList(): List<T> {
    return Pager(PagingConfig(pageSize = 10), pagingSourceFactory = { this }).flow.asSnapshot {
        scrollTo(50)
    }
}
```

Reply:
Thank you for pointing that out, my implementation felt quite hacky indeed.